xla/service/gpu/cancel_all_gather_dynamic_slice.cc

```diff
@@ -0,0 +1,62 @@
+/* Copyright 2024 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "xla/service/gpu/cancel_all_gather_dynamic_slice.h"
+
+#include "xla/hlo/ir/hlo_casting_utils.h"
+#include "xla/service/collective_opt_utils.h"
+
+namespace xla {
+bool CancelAllGatherDynamicSlice::InstructionMatchesPattern(
+    HloInstruction* instruction) {
+  if (instruction->opcode() != HloOpcode::kDynamicSlice) {
+    return false;
+  }
+  // Two cases: dynamic-slice(reshape(all-gather)) or dynamic-slice(all-gather)
```

Comment:
You can cast `instruction` to HloDynamicSliceInstruction and use the named accessors. The comment helps but it'd be nice if the condition was easier to read. Maybe extract some local variables.